semantic-python/src/Language/Python/Core.hs

```diff
@@ -32,21 +35,44 @@ type CoreSyntax sig t = ( Member Core sig
 
 class Compile py where
   -- FIXME: we should really try not to fail
-  compile :: (CoreSyntax sig t, MonadFail m) => py -> m (t Name)
+  compile :: ( CoreSyntax syn t
+             , Member (Reader Path.RelFile) sig
+             , Carrier sig m
+             , MonadFail m
+             )
+          => py
+          -> m (t Name)
 
   default compile :: (MonadFail m, Show py) => py -> m (t Name)
   compile = defaultCompile
 
-  compileCC :: (CoreSyntax sig t, MonadFail m) => py -> m (t Name) -> m (t Name)
-
-  default compileCC :: (CoreSyntax sig t, MonadFail m) => py -> m (t Name) -> m (t Name)
+  compileCC :: ( CoreSyntax syn t
+               , Member (Reader Path.RelFile) sig
+               , Carrier sig m
+               , MonadFail m
+               )
+            => py
+            -> m (t Name)
+            -> m (t Name)
+
+  default compileCC :: ( CoreSyntax syn t
+                       , Member (Reader Path.RelFile) sig
+                       , Carrier sig m
+                       , MonadFail m
+                       ) => py -> m (t Name) -> m (t Name)
   compileCC py cc = (>>>) <$> compile py <*> cc
 
-locate :: (HasField "ann" syntax Span, CoreSyntax sig t) => syntax -> t a -> t a
-locate syn = Core.annAt (locFromTSSpan (getField @"ann" syn))
-  where
-    locFromTSSpan (TreeSitter.Span (TreeSitter.Pos a b) (TreeSitter.Pos c d))
-      = Data.Loc.Loc mempty (Data.Loc.Span (Data.Loc.Pos a b) (Data.Loc.Pos c d))
+locate :: ( HasField "ann" syntax Span
+          , CoreSyntax syn t
+          , Member (Reader Path.RelFile) sig
+          , Carrier sig m
+          ) => syntax -> t a -> m (t a)
+locate syn item = do
+  fp <- asks @Path.RelFile (Text.pack . Path.toString)
```

Comment:
Core’s motivation for having this information is twofold:

1. Implementing `__FILE__` and similar directives (not yet implemented).
2. Debugging support; “where did this come from?” or “where am I in the program?”

Both of those require the thing to be represented as text, so ultimately using `Path.RelFile` means we’re either converting the thing to `Text` per-node like this, or else per-request, and neither of those seems ideal—tho of the two, I agree that I’d prefer the latter. I think this implies that we actually want to have a `Reader Text` effect instead of `Reader Path.RelFile`.

I realize this is what you were trying to avoid in the first place, but I honestly think that `Path.RelFile` needs to occur further upstream for us: whatever we end up using to manipulate paths in the first place. By the time it ends up in core, we’re not going to be using it to actually load or build new paths (I think, anyway), so the extra safety it gives us is not especially valuable to us.

In the meantime, I’ve no objection to leaving this in place for the moment since it’s working and all, but we should probably consider the efficiency of this as we eventually start to abstract parts of `semantic-python` into either `semantic-core` or a `semantic-compiler` (`semanticc`?) package.

Reply:
I changed this from a `Path.RelFile` to a newtype over `Text`, which both announces more explicitly this datum’s purpose and avoids repeated inefficient copies.